### PR TITLE
[GlobalIsel][AArch64] legalize ptr add

### PR DESCRIPTION
**Pull request by tschuett**
LLVM ERROR: unable to legalize instruction: %275:_(<4 x p0>) = G_PTR_ADD %268:_, %274:_(<4 x s64>) (in function: prepare_for_pass)

### CODE REVIEW

**Comment by tschuett** [comment]
3 x p0 asserts

**Comment by tschuett** [comment]
```
Assertion failed: (llvm::all_of(SrcOps, [&, this](const SrcOp &Op) { return Op.getLLTTy(*getMRI()) == SrcOps[0].getLLTTy(*getMRI()); }) && "type mismatch in input list"), function buildInstr, file MachineIRBuilder.cpp, line 1355.
```

**Comment by dc03-work** [comment]
> 3 x p0 asserts

Does your patch fix this or is it still the case after applying your patch?

**Comment by dc03-work** [comment]
> After this patch, it still asserts.

You can likely put a `moreElementsToNextPow2` before the `clampNumElements` to address this.

File: llvm/lib/Target/AArch64/GISel/AArch64LegalizerInfo.cpp [hunk 1]
```diff
@@ -183,7 +183,9 @@ AArch64LegalizerInfo::AArch64LegalizerInfo(const AArch64Subtarget &ST)
 
   getActionDefinitionsBuilder(G_PTR_ADD)
       .legalFor({{p0, s64}, {v2p0, v2s64}})
-      .clampScalar(1, s64, s64);
+      .clampScalar(1, s64, s64)
+      .clampNumElements(0, v2p0, v2p0)
+      .clampNumElements(1, v2s64, v2s64);
```
[the remark this answers is not in the document]
Comment (tschuett) [reply llvm/lib/Target/AArch64/GISel/AArch64LegalizerInfo.cpp:188]: // TODO: Is the offset allowed to be a scalar with a vector?

I would argue for safeness despite redundancy.

[the remark this answers is not in the document]
Comment (tschuett) [reply llvm/lib/Target/AArch64/GISel/AArch64LegalizerInfo.cpp:188]: I vote for both: we get fewer and more elements for both type indices.

**Comment by tschuett** [comment]
Ping.

**Comment by dc03-work** [comment]
> The new error after `moreElementsToNextPow2` is:
> 
> ```
> unable to legalize instruction: %res:_(<3 x p0>) = G_PTR_ADD %ptr:_, %add:_(<3 x s64>) (in function: test_ptr_add_vec_3xp0)
> ```

Do you have a more detailed log? I guess scalarization can be done for odd-sized vectors, but having moreElements support would be a better solution.

**Comment by tschuett** [comment]
I want to address the reported issue.

**Comment by tschuett** [comment]
I only got the error message about legalization..

**Comment by tschuett** [comment]
Unfortunately, I cannot add a ll file there are more issues:
```
bb.1 (%ir-block.0):
  liveins: $q0, $x0
  %0:_(p0) = COPY $x0
  %1:_(<4 x s32>) = COPY $q0
  %2:_(<4 x p0>) = G_BUILD_VECTOR %0:_(p0), %0:_(p0), %0:_(p0), %0:_(p0)
  %3:_(<4 x s64>) = G_SEXT %1:_(<4 x s32>)
  %4:_(<4 x p0>) = G_PTR_ADD %2:_, %3:_(<4 x s64>)
  %5:_(<4 x p0>) = COPY %4:_(<4 x p0>)
  %6:_(<2 x s64>), %7:_(<2 x s64>) = G_UNMERGE_VALUES %5:_(<4 x p0>)
  $q0 = COPY %6:_(<2 x s64>)
  $q1 = COPY %7:_(<2 x s64>)
  RET_ReallyLR implicit $q0, implicit $q1

# End machine code for function ptr_add_4_vector_ptr.

*** Bad machine code: G_UNMERGE_VALUES source operand does not match vector destination operands ***
```

**Comment by tschuett** [comment]
Test:
```
define <4 x ptr> @ptr_add_4_vector_ptr(ptr %vec, <4 x i32> %offset) {
  %ptr_add = getelementptr i8, ptr %vec, <4 x i32> %offset
  ret <4 x ptr> %ptr_add
}
```

**Comment by tschuett** [comment]
opened issue for the crash
https://github.com/llvm/llvm-project/issues/90202

**Comment by davemgreen** [comment]
Thanks - it looks like that G_UNMERGE_VALUES is maybe invalid, and it is probably the return type that is going wrong.

Can you change the test to store a value to show this is working now, and we can look at #90202 separately.

**Comment by tschuett** [comment]
What do you mean by store? The mir test shows that it successfully legalized the ptradd.

**Comment by davemgreen** [comment]
> What do you mean by store? The mir test shows that it successfully legalized the ptradd.

```
define void @ptr_add_4_vector_ptr(ptr %vec, <4 x i32> %offset) {
  %ptr_add = getelementptr i8, ptr %vec, <4 x i32> %offset
  store <4 x ptr> %ptr_add, ptr %vec
  ret void
}
```

As the problems you have encountered has shown, mir tests are not very useful in isolation and end-to-end tests work better if they are available.

**Comment by tschuett** [comment]
The error message at the top is about legalization. The error is about calling conventions. It is unrelated.

File: llvm/lib/Target/AArch64/GISel/AArch64LegalizerInfo.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (arsenm) [reply llvm/lib/Target/AArch64/GISel/AArch64LegalizerInfo.cpp:188]: Drop the second one - it's adding untested paths

**Review by arsenm — APPROVED** [review]
LGTM but I'd prefer to drop the untested clamp source type

**Pull request merged**